### PR TITLE
[Track2 Function #1] Add entities for function in app service library

### PR DESCRIPTION
**Pull request by Flanker32**
Add entities for function in app service library

### CODE REVIEW

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java [hunk 1]
```diff
@@ -46,10 +52,16 @@
         return values;
     }
 
+    public static List<PricingTier> valuesForWebApps() {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java:55]: forWebapps()

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java [hunk 2]
```diff
@@ -46,10 +52,16 @@
...
+        return values().stream()
+                .filter(pricing -> !StringUtils.equalsAnyIgnoreCase(pricing.tier, "Dynamic", "ElasticPremium"))
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java:57]: extract method: is function pricing tier

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java [hunk 3]
```diff
@@ -64,6 +64,10 @@ public static PricingTier fromString(String input) {
                 .findFirst().orElse(null);
     }
 
+    public static boolean isValidWebAppPricingTier(PricingTier pricingTier) {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java:67]: isFunctionPricingTier

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java [hunk 4]
```diff
@@ -64,6 +64,10 @@ public static PricingTier fromString(String input) {
...
+        return !StringUtils.equalsAnyIgnoreCase(pricingTier.getTier(), "Dynamic", "ElasticPremium");
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java:68]: !StringUtils.equalsAnyIgnoreCase -> StringUtils.equalsAnyIgnoreCase

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java [hunk 5]
```diff
@@ -52,9 +52,9 @@
         return values;
     }
 
-    public static List<PricingTier> valuesForWebApps() {
+    public static List<PricingTier> forWebapps() {
         return values().stream()
-                .filter(pricing -> !StringUtils.equalsAnyIgnoreCase(pricing.tier, "Dynamic", "ElasticPremium"))
+                .filter(PricingTier::isValidWebAppPricingTier)
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java:57]: !isFunctionPricingTier

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/FunctionDeployType.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+import com.microsoft.azure.toolkit.lib.common.exception.AzureExecutionException;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.Arrays;
+
+public enum FunctionDeployType {
+    FTP,
+    ZIP,
+    EMPTY,
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/FunctionDeployType.java:16]: what is this used for?

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/FunctionDeployType.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    private static final String UNKNOWN_DEPLOYMENT_TYPE = "The value of <deploymentType> is unknown.";
+
+    public static FunctionDeployType fromString(final String input) throws AzureExecutionException {
+        if (StringUtils.isEmpty(input)) {
+            return EMPTY;
+        }
+        return Arrays.stream(FunctionDeployType.values())
+                .filter(type -> StringUtils.equalsAnyIgnoreCase(type.name(), input))
+                .findFirst().orElseThrow(() -> new AzureExecutionException(UNKNOWN_DEPLOYMENT_TYPE));
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/FunctionDeployType.java:29]: use `AzureToolkitRuntimeException` instead if there are no dedicated exception. we would deprecate `AzureExecutionException`.

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java [hunk 8]
```diff
@@ -46,10 +52,20 @@
         return values;
     }
 
+    public static List<PricingTier> forWebapps() {
+        return values().stream()
+                .filter(pricingTier -> !isExclusivePricingTierForFunctionApp(pricingTier))
+                .collect(Collectors.toList());
+    }
+
     public static PricingTier fromString(String input) {
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java:61]: `fromSize()` or `fromString(String size)`?

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java [hunk 9]
```diff
@@ -46,10 +52,20 @@
         return values;
     }
 
+    public static List<PricingTier> forWebapps() {
+        return values().stream()
+                .filter(pricingTier -> !isExclusivePricingTierForFunctionApp(pricingTier))
+                .collect(Collectors.toList());
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/PricingTier.java:58]: I would prefer a static list. which can be much faster.

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/FunctionDeployType.java (hunk 6 above)
Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/model/FunctionDeployType.java:16]: Deploy methods for function, like web app one deploy

**Review by wangmingliang-ms — APPROVED** [review]

**Pull request merged**